tools/wpt/virtualenv.py

```diff
@@ -4,7 +4,17 @@
 import logging
 from distutils.spawn import find_executable
 
-import pkg_resources
+# The `pkg_resources` module is provided by `virtualenv`. Tolerate its absence
+# so that this module may be evaluated when that module is not available.
+# Because users may not recognize the `pkg_resources` module by name, raise a
+# more descriptive error if it is referenced during execution.
+try:
+    import pkg_resources
+except ImportError:
+    class Pkg_resources(object):
```

Comment:
Let's not do this like this, let's just make `working_set` raise (uh, something) if it doesn't get imported, and handle that case in `install`/`install_requirements` by unconditionally calling `pip`.

Reply:
The module is also referenced by the `install` and `install_requirements`
method. That highlights my motivation for implementing it in this way: it
alleviates us of the responsibility of maintaining more descriptive error
messages in multiple places (and avoids causing confusion when we inevitably
forget).

I can appreciate your general unease with this pattern, though. A function that
encapsulates this would let us defer the import and consistently report an
error:

```python
def get_pkg_resources():
    try:
        import pkg_resources
        return pkg_resources
    except ImportError:
        raise ValueError("The Python module `virtualenv` is not installed.")
```

What do you say?

Reply:
I have no strong opinion; I think I was expecting the references in `install`/`install_requirements` to be behind a branch when `working_set` can be returned. But, eh, do whatever!

I'd rather the import be at the top level, though, rather than in a function. Much easier for static analysis tools to keep track of things.

Reply:
> I'd rather the import be at the top level, though, rather than in a function. Much easier for static analysis tools to keep track of things.

Strongly agree. This patch now unconditionally imports the module and instead branches on the definition of the function.